File: src/cmds/docker.ts
import { deduplicateLines } from '../core/filter';
import { run } from '../core/runner';
import { stripAnsi } from '../core/utils';
import { HandlerResult } from './git';

export function handleDocker(args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || '';
  const result = run('docker', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  let cmdType = `docker ${sub || 'cmd'}`;

  try {
    if (sub === 'ps') filtered = formatPs(raw, ultra);
    else if (sub === 'images') filtered = formatImages(raw, ultra);
    else if (sub === 'logs') filtered = formatLogs(raw, ultra);
    else if (sub === 'compose') filtered = formatCompose(raw, ultra);
    else filtered = stripAnsi(raw).trim();
  } catch {
    filtered = raw;
  }

  return {
    filteredOutput: filtered || (result.exitCode === 0 ? 'ok' : raw),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType,
    execMs: result.execMs,
  };
}

export function handleKubectl(args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || '';
  const result = run('kubectl', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';

  try {
    if (sub === 'logs') {
      filtered = formatLogs(raw, ultra);
    } else if (sub === 'get') {
      filtered = formatKubectlGet(raw, ultra);
    } else {
      filtered = stripAnsi(raw).trim();
    }
  } catch {
    filtered = raw;
  }

  return {
    filteredOutput: filtered || (result.exitCode === 0 ? 'ok' : raw),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType: `kubectl ${sub || 'cmd'}`,
    execMs: result.execMs,
  };
}

function formatPs(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n').filter((l) => l.trim());
  const data = lines.slice(1);
  const running = data.filter((l) => /\bUp\b/.test(l)).length;
  const stopped = data.length - running;
  if (ultra) return `${running}↑/${stopped}↓`;
  if (data.length === 0) return 'no containers';
  const out = [`${data.length} container${data.length === 1 ? '' : 's'}: ${running} running, ${stopped} stopped`];
  for (const line of data.slice(0, 10)) {
    const tokens = line.split(/\s{2,}/);
    if (tokens.length >= 2) {
      const id = tokens[0].slice(0, 12);
      const image = tokens[1];
      const status = tokens.find((t) => /\b(Up|Exited|Created)\b/.test(t)) || '';
      out.push(`  ${id} ${image} ${status}`);
    }
  }
  return out.join('\n');
}

function formatImages(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n').filter((l) => l.trim());
  const data = lines.slice(1);
  if (ultra) return `${data.length} images`;
  if (data.length === 0) return 'no images';
  const out = [`${data.length} image${data.length === 1 ? '' : 's'}:`];
  for (const line of data.slice(0, 15)) {
    const tokens = line.split(/\s{2,}/);
    if (tokens.length >= 2) {
      out.push(`  ${tokens[0]}:${tokens[1] || 'latest'}`);
    }
  }
  return out.join('\n');
}

function formatLogs(raw: string, ultra: boolean): string {
  const dedup = deduplicateLines(raw);
  const lines = dedup.split('\n').filter((l) => l.trim());
  const errors = lines.filter((l) => /\b(error|err|failed|exception|panic)\b/i.test(l));
  if (ultra) {
    const top = lines.slice(0, 3).map((l) => l.slice(0, 60)).join(' | ');
    return `${lines.length}L ${errors.length}E | ${top}`;
  }
  if (lines.length === 0) return 'no logs';
  const out = [`${lines.length} unique log line${lines.length === 1 ? '' : 's'} (${errors.length} errors)`];
  if (errors.length > 0) {
    out.push('');
    out.push('Errors:');
    for (const e of errors.slice(0, 10)) out.push(`  ${e}`);
  }
  out.push('');
  out.push('Top messages:');
  for (const l of lines.slice(0, 10)) out.push(`  ${l}`);
  return out.join('\n');
}

function formatCompose(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n').filter((l) => l.trim());
  if (ultra) return `compose: ${lines.length}L`;
  return lines.slice(-30).join('\n');
}

function formatKubectlGet(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n').filter((l) => l.trim());
  const data = lines.slice(1);
  if (ultra) return `${data.length} resources`;
  return `${data.length} resource${data.length === 1 ? '' : 's'}\n${data.slice(0, 20).join('\n')}`;
}

File: internal/handlers/docker.go
package handlers

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/filter"
	"github.com/Kevalgor12/tok-proxy/internal/run"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

var (
	twoSpaceRe   = regexp.MustCompile(`\s{2,}`)
	dockerUpRe   = regexp.MustCompile(`\bUp\b`)
	dockerStatRe = regexp.MustCompile(`\b(Up|Exited|Created)\b`)
	logErrRe     = regexp.MustCompile(`(?i)\b(error|err|failed|exception|panic)\b`)
)

func HandleDocker(args []string, ultra bool) Result {
	sub := ""
	if len(args) > 0 {
		sub = args[0]
	}
	r := run.Run("docker", args)
	raw := combined(r)
	var filtered string

	switch sub {
	case "ps":
		filtered = formatPs(raw, ultra)
	case "images":
		filtered = formatImages(raw, ultra)
	case "logs":
		filtered = formatLogs(raw, ultra)
	case "compose":
		filtered = formatCompose(raw, ultra)
	default:
		filtered = strings.TrimSpace(util.StripAnsi(raw))
	}

	cmdType := "docker " + firstNonEmpty(sub, "cmd")
	return finalizeRaw(filtered, r, raw, cmdType)
}

func HandleKubectl(args []string, ultra bool) Result {
	sub := ""
	if len(args) > 0 {
		sub = args[0]
	}
	r := run.Run("kubectl", args)
	raw := combined(r)
	var filtered string

	switch sub {
	case "logs":
		filtered = formatLogs(raw, ultra)
	case "get":
		filtered = formatKubectlGet(raw, ultra)
	default:
		filtered = strings.TrimSpace(util.StripAnsi(raw))
	}

	return finalizeRaw(filtered, r, raw, "kubectl "+firstNonEmpty(sub, "cmd"))
}

func formatPs(raw string, ultra bool) string {
	data := dataRows(util.StripAnsi(raw))
	running := 0
	for _, l := range data {
		if dockerUpRe.MatchString(l) {
			running++
		}
	}
	stopped := len(data) - running
	if ultra {
		return fmt.Sprintf("%d↑/%d↓", running, stopped)
	}
	if len(data) == 0 {
		return "no containers"
	}
	out := []string{fmt.Sprintf("%d %s: %d running, %d stopped",
		len(data), plural(len(data), "container", "containers"), running, stopped)}
	for _, line := range limit(data, 10) {
		tokens := twoSpaceRe.Split(line, -1)
		if len(tokens) >= 2 {
			id := tokens[0]
			if len(id) > 12 {
				id = id[:12]
			}
			status := ""
			for _, t := range tokens {
				if dockerStatRe.MatchString(t) {
					status = t
					break
				}
			}
			out = append(out, fmt.Sprintf("  %s %s %s", id, tokens[1], status))
		}
	}
	return strings.Join(out, "\n")
}

func formatImages(raw string, ultra bool) string {
	data := dataRows(util.StripAnsi(raw))
	if ultra {
		return fmt.Sprintf("%d images", len(data))
	}
	if len(data) == 0 {
		return "no images"
	}
	out := []string{fmt.Sprintf("%d %s:", len(data), plural(len(data), "image", "images"))}
	for _, line := range limit(data, 15) {
		tokens := twoSpaceRe.Split(line, -1)
		if len(tokens) >= 2 {
			tag := tokens[1]
			if tag == "" {
				tag = "latest"
			}
			out = append(out, fmt.Sprintf("  %s:%s", tokens[0], tag))
		}
	}
	return strings.Join(out, "\n")
}

func formatLogs(raw string, ultra bool) string {
	lines := nonEmptyLines(filter.DeduplicateLines(raw))
	var errors []string
	for _, l := range lines {
		if logErrRe.MatchString(l) {
			errors = append(errors, l)
		}
	}
	if ultra {
		var top []string
		for _, l := range limit(lines, 3) {
			top = append(top, clip(l, 60))
		}
		return fmt.Sprintf("%dL %dE | %s", len(lines), len(errors), strings.Join(top, " | "))
	}
	if len(lines) == 0 {
		return "no logs"
	}
	out := []string{fmt.Sprintf("%d unique %s (%d errors)",
		len(lines), plural(len(lines), "log line", "log lines"), len(errors))}
	if len(errors) > 0 {
		out = append(out, "", "Errors:")
		for _, e := range limit(errors, 10) {
			out = append(out, "  "+e)
		}
	}
	out = append(out, "", "Top messages:")
	for _, l := range limit(lines, 10) {
		out = append(out, "  "+l)
	}
	return strings.Join(out, "\n")
}

func formatCompose(raw string, ultra bool) string {
	lines := nonEmptyLines(util.StripAnsi(raw))
	if ultra {
		return fmt.Sprintf("compose: %dL", len(lines))
	}
	return strings.Join(lastN(lines, 30), "\n")
}

func formatKubectlGet(raw string, ultra bool) string {
	data := dataRows(util.StripAnsi(raw))
	if ultra {
		return fmt.Sprintf("%d resources", len(data))
	}
	return fmt.Sprintf("%d %s\n%s", len(data), plural(len(data), "resource", "resources"),
		strings.Join(limit(data, 20), "\n"))
}

// dataRows returns the non-empty lines of a table with the header row dropped.
func dataRows(clean string) []string {
	lines := nonEmptyLines(clean)
	if len(lines) <= 1 {
		return nil
	}
	return lines[1:]
}

func nonEmptyLines(s string) []string {
	var out []string
	for _, l := range strings.Split(s, "\n") {
		if strings.TrimSpace(l) != "" {
			out = append(out, l)
		}
	}
	return out
}

func clip(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}
